### ISSUE TITLE
eliminate RewindableProperty

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/gravity-and-orbits#276]
As part of https://github.com/phetsims/axon/issues/213, all Property subtypes (including RewindableProperty) were converted to classes.   RewindableProperty overrides `set`, and the additional code is equivalent to a listener that is called _after_ all other listeners are notified.

Please explain:

(1) ... why GRAPHING_LINES/GamePhaseProperty was flagged as unacceptable (see https://github.com/phetsims/graphing-lines/issues/109), but no similar issue was created for RewindableProperty.

(2) ... why RewindableProperty doesn't represent a complication for PhET-iO instrumentation.  It is much more complicated than GamePhaseProperty.

(3) ... why RewindableProperty shouldn't be replaced with a solution that doesn't involve subclassing Property.

**Comment by pixelzoom** [comment]
But in https://github.com/phetsims/graphing-lines/issues/109#issuecomment-455947852, @samreid said:
> Extending Property to add something that is like a listener but not a listener seems odd.

Here's RewindableProperty `set`:

```js
    /**
     * @public
     * @override
     */
    set( value ) {
      super.set( value );

      // If the user changed the initial conditions (as opposed to the state changing through model stepping),
      // then store the new initial conditions, which can be rewound to
      if ( this.changeRewindValueProperty.get() ) {
        this.storeRewindValueNoNotify();

        for ( let i = 0; i < this.rewindValueChangedListeners.length; i++ ) {
          this.rewindValueChangedListeners[ i ]();
        }
      }
      this.differentProperty.set( !this.equalsRewindPoint() );
    }
```

This looks perfectly legitimate to me not at all "odd".  But it's essentially using the same pattern that was used in GamePhaseProperty, except that what's been added to `set` here is equivalent to a listener that is called _after_ all other listeners have been notified.

**Comment by samreid** [comment]
> (1) ... why GRAPHING_LINES/GamePhaseProperty was flagged as unacceptable (see phetsims/graphing-lines#109), but no similar issue was created for RewindableProperty.

GamePhaseProperty was never declared unacceptable.  I think it is a good policy that when a developer sees something odd or confusing, they should point it out or bring it up for discussion.  That is exactly what happened.  Do you disagree with this policy?

> (2) ... why RewindableProperty doesn't represent a complication for PhET-iO instrumentation. 

Yes, it is a complex feature and it will require additional effort for PhET-iO instrumentation.  I expect we will need to introduce a new IO type for RewindableProperty.  There may be an alternative/better way to accomplish this, but I can't think of one right now.  I recommend to leave this issue open for when this sim is revisited.

> why RewindableProperty shouldn't be replaced with a solution that doesn't involve subclassing Property.

We both agree it is a great idea to subclass Property where appropriate.  The difference between RewindableProperty and GamePhaseProperty is that GamePhaseProperty needs to have something happen first, and hence could technically use listeners[0].  RewindProperty has to have something last, and hence cannot add a listener at a specific point in the listener array, since listeners may be added at any time.

> and certainly preferable to either adding specific (unnecessary) support to Property or adding listeners in a specific order.

Property listeners are invoked in the order they were added.  Same for Emitter.  I agree AXON doesn't need further support for specifying listener ordering.

It seems like the the two competing perspectives are (a) `hook` is a non-listener function that should be called before the Property listeners are notified or (b) `hook` is a Property listener that should be called before other listeners.  When I came across this code it seemed like (b), but I now understand that you firmly believe (a).  That is OK with me!

As I said above, I recommend we leave this issue open until the sim is revisited.  I would also prefer that further discussion on this issue or https://github.com/phetsims/graphing-lines/issues/109 be done over voice or in person.

**Comment by samreid** [comment]
I reviewed RewindableProperty, simplified and cleaned it up a bit, and it seems like it provides the feature needed by Gravity and Orbits of being able to store and restore a specific value (in addition to the overall reset value).  @pixelzoom would you recommend any further work or discussion before this issue can be closed?

**Comment by pixelzoom** [comment]
Changes look good, looks like most removing stuff that wasn't used.  

I'm OK to close this.  But can you summarize where we ended up on this, and how it differs from https://github.com/phetsims/graphing-lines/issues/109?

**Comment by samreid** [comment]
Gravity and Orbits requires the ability to store and restore specific states that differ from the overall reset value.  This is accomplished via RewindableProperty.  GraphingLines needs to make other state changes before `gamePhaseProperty`'s value changes, this is accomplished by performing that work in `BaseGameModel.setGamePhase` before setting the Property.  Please reopen if there's more to discuss.
